scripts/release/main.go

```diff
@@ -221,3 +275,114 @@ func removeMainlineBlurb(body string) string {
 
 	return strings.Join(newBody, "\n")
 }
+
+// autoversion automatically updates the provided channel to version in markdown
+// files.
+func (r *releaseCommand) autoversion(ctx context.Context, channel, version string) error {
+	var files []string
+
+	// For now, scope this to docs, perhaps we include README.md in the future.
+	if err := afero.Walk(r.fs, "docs", func(path string, _ fs.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if strings.EqualFold(filepath.Ext(path), ".md") {
+			files = append(files, path)
+		}
+		return nil
+	}); err != nil {
+		return xerrors.Errorf("walk failed: %w", err)
+	}
+
+	for _, file := range files {
+		err := r.autoversionFile(ctx, file, channel, version)
+		if err != nil {
+			return xerrors.Errorf("autoversion file failed: %w", err)
+		}
+	}
+
+	return nil
+}
+
+// autoversionMarkdownPragmaRe matches the autoversion pragma in markdown files.
+//
+// Example:
+//
+//	<!-- autoversion(stable): "--version [version]"" -->
+//
+// The channel is the first capture group and the match string is the second
+// capture group. The string "[version]" is replaced with the new version.
+var autoversionMarkdownPragmaRe = regexp.MustCompile(`<!-- ?autoversion\(([^)]+)\): ?"([^"]+)" ?-->`)
+
+func (r *releaseCommand) autoversionFile(ctx context.Context, file, channel, version string) error {
+	version = strings.TrimPrefix(version, "v")
+	logger := r.logger.With(slog.F("file", file), slog.F("channel", channel), slog.F("version", version))
+
+	logger.Debug(ctx, "checking file for autoversion pragma")
+
+	contents, err := afero.ReadFile(r.fs, file)
+	if err != nil {
+		return xerrors.Errorf("read file failed: %w", err)
+	}
+
+	lines := strings.Split(string(contents), "\n")
+	var matchRe *regexp.Regexp
+	for i, line := range lines {
+		if autoversionMarkdownPragmaRe.MatchString(line) {
+			matches := autoversionMarkdownPragmaRe.FindStringSubmatch(line)
+			matchChannel := matches[1]
+			match := matches[2]
+
+			logger := logger.With(slog.F("line_number", i+1), slog.F("match_channel", matchChannel), slog.F("match", match))
+
+			logger.Debug(ctx, "autoversion pragma detected")
+
+			if matchChannel != channel {
+				logger.Debug(ctx, "channel mismatch, skipping")
+				continue
+			}
+
+			logger.Info(ctx, "autoversion pragma found with channel match")
+
+			match = strings.Replace(match, "[version]", `(?P<version>[0-9]+\.[0-9]+\.[0-9]+)`, 1)
+			logger.Debug(ctx, "compiling match regexp", "match", match)
+			matchRe, err = regexp.Compile(match)
```

Comment:
maybe move it above and just do MustCompile (shorter code)?

Reply:
I think a panic here could be jarring. Since we’re actually compiling the contents from the markdown file, there’s no way to do this ahead of time :/